### DISCUSSION TITLE
Meta tags are not generated when creating SSG pages with Dynamic Routes.

### DISCUSSION THREAD

**Discussion by darrylmorley** [vercel/next.js#21470] [Help] [answered]
Hi,

Meta tags are not generated when creating SSG pages with Dynamic Routes, using 'next/head'.  What is the recommended method of using dynamic data in the head of dynamic pages?

**Comment by michrome** [comment]
If you're wanting **dynamic** meta data, your might be better of with `getServerSideProps`. Static Site Generation (SSG) is typically used with static data because the page is built during the build, not at request time.

**Reply by darrylmorley** [reply]
Hi thanks for responding,

I'm trying to populate meta data with the data from props on static pages, maybe dynamic meta is not what I'm looking for.

**Reply by michrome** [reply]
You could do something along the lines of this:

```javascript
// components/layout.js
import Head from "next/head";

export default function Layout({ metaDescription, children }) {
  return (
    <>
      <Head>
        <meta name="description" content={metaDescription} />
      </Head>
      {children}
    </>
  );
}

```

```javascript
//pages/my-page.js
import Layout from "../components/layout";

export default function MyPage() {
  const metaDescription = "A description"; // Set this from props
  return <Layout metaDescription={metaDescription}>Content here</Layout>;
}
```

**Reply by darrylmorley** [reply]
I think this is not too dissimilar to what I'm already doing, but I'll give it a try and update. Thanks

**Reply by darrylmorley** [reply]
As I expected, I'm getting the same result. So, maybe I'm expecting something that I shouldn't?

Meta is not visible in the page source but is visible via the inspector. Some spiders do not see the meta content , including the facebook sharing debugger, but Google Search Console appears to recognise it. Is this as expected or is it possible to have the meta in the page source?

**Reply by michrome** [reply]
@darrylmorley is https://j4e.name doing what you're looking for? If so, I am happy to share the code.

**Reply by darrylmorley** [reply]
@michrome Yes that's what I'm looking for 👍

**Reply by michrome** [reply]
```javascript
// pages/index.js
import Link from "next/link";
import { articlesWithMetadata } from "../lib/content";
import ArticleLayout from "../components/article-layout";

const meta = {
  datePublished: "2020-06-07",
  description: "Mike Jerome’s website",
  title: "j4e.name",
};

function HomePage({ articles, meta }) {
  return (
    <ArticleLayout meta={meta}>
      <dl>
        {articles.map((article) => (
          <div key={article.slug} className="mb-6">
            <dt>
              <Link href={article.path}>
                <a>{article.meta.title}</a>
              </Link>
            </dt>
            <dd>{article.meta.description}</dd>
          </div>
        ))}
      </dl>
    </ArticleLayout>
  );
}

export async function getStaticProps(context) {
  let articles = [];
  try {
    articles = await articlesWithMetadata();
  } catch (error) {
    console.log(error);
  }
  return {
    props: { articles: articles, meta: metaWithModifiedFromArticles(articles) },
  };
}

export function metaWithModifiedFromArticles(articles) {
  meta.dateModified = articles[0].meta.datePublished;
  return meta;
}

export default HomePage;
```


```javascript
// components/article-layout.js
import Head from "next/head";
import Link from "next/link";
import { useRouter } from "next/router";
import { canonicalUrl } from "../lib/helpers";
import SchemaArticle from "./schema-article";

export default function ArticleLayout({ meta, children }) {
  const router = useRouter();
  meta.canonicalUrl = canonicalUrl(router.pathname);
  return (
    <>
      <Head>
        <title>{meta.title}</title>
        {meta.description && (
          <meta name="description" content={meta.description} />
        )}
        <SchemaArticle meta={meta} />
        <link rel="canonical" href={meta.canonicalUrl} />
      </Head>
      <nav className="md:fixed md:transform-gpu md:origin-top-left md:rotate-90 md:-translate-x-6 md:translate-y-6">
        <div className="md:transform-gpu md:rotate-180">
          <Link href="/">
            <a>j4e.name</a>
          </Link>
        </div>
      </nav>
      <article>
        <h1>{meta.title}</h1>
        {children}
      </article>
    </>
  );
}
```

**Reply by darrylmorley** [reply]
Thank you for sharing. I'll have a gander.

**Reply by michrome** [reply]
No problem. My suspicion is you're not using `getStaticProps`.

**Reply by darrylmorley** [reply]
I'm using getStaticPaths and getStaticProps.

**Reply by darrylmorley** [reply]
I've tried every which way, It doesn't appear to be possible to render meta on build from dynamic data. It is only rendered on the client side. Quite an issue.

**Reply by michrome** [reply]
> I've tried every which way, It doesn't appear to be possible to render meta on build from dynamic data. It is only rendered on the client side. Quite an issue.

It’s absolutely possible, as I’ve shown. Do you have code or a repo to share?

**Reply by darrylmorley** [reply]
It all works fine until getStaticPaths is used by the looks of things. If you don't mind having a look the repo is https://github.com/Shooting-Supplies-Ltd/fab-lsretail.

**Comment by darrylmorley** [comment, marked as answer]
So, if you go to https://fabdefense.co.uk and view source you can see the meta tags, if you go to the shops page, you can again see the meta tags completed. 

But, if you then select a product and view source, no meta tags, if you inspect the page then tags are in the head section on render. 

It appears that next is not building the products as static pages as I would expect from getStaticProps. I'm not sure what to make of that or how to start resolving it.

**Reply by michrome** [reply]
I had a _quick_ look. Wondering if it's because you have a static prop named `Item` (capital I) but your meta tags are driven by `item` (lowercase i) which is initialised with `const [item, setItem] = useState()`.

**Reply by darrylmorley** [reply]
You know what, I think you might be right 😒. I'll update and let you know, thanks 👍

**Reply by darrylmorley** [reply]
Thanks @michrome!
